overlord/devicestate/handlers_systems.go

```diff
@@ -0,0 +1,307 @@
+	if err != nil {
+		return fmt.Errorf("cannot create a recovery system with label %q for %v: %v", label, model.Model(), err)
+	}
+	logger.Debugf("recovery system dir: %v", systemDir)
+	logger.Debugf("new common snap files: %v", newFiles)
+
+	defer func() {
+		if err == nil {
+			return
+		}
+		if err := os.RemoveAll(systemDir); err != nil && !os.IsNotExist(err) {
+			logger.Noticef("when removing recovery system %q: %v", label, err)
+		}
+		for _, f := range newFiles {
+			// new files under the recovery system dir would have
+			// been removed with the directory already
+			if err := os.Remove(f); err != nil && !os.IsNotExist(err) {
+				logger.Noticef("when removing seed file %q: %v", f, err)
+			}
+		}
+	}()
+
+	setup.NewCommonFiles = newFiles
+	setup.Directory = systemDir
+	// 2. keep track of the system in task state
+	if err := setTaskRecoverySystemSetup(t, setup); err != nil {
+		return fmt.Errorf("cannot record recovery system setup state: %v", err)
+	}
+	// 3. set up boot variables for tracking the tried system state
+	if err := boot.SetTryRecoverySystem(remodelCtx, label); err != nil {
+		// rollback?
+		return fmt.Errorf("cannot attempt booting into recovery system %q: %v", label, err)
+	}
+	// 4. and set up the next boot that that system
+	if err := boot.SetRecoveryBootSystemAndMode(remodelCtx, label, "recover"); err != nil {
+		return fmt.Errorf("cannot set device to boot into candidate system %q: %v", label, err)
+	}
+
+	// this task is done, further processing happens in finalize
+	t.SetStatus(state.DoneStatus)
+
+	logger.Noticef("restarting into candidate system %q", label)
+	m.state.RequestRestart(state.RestartSystemNow)
+	return nil
+}
+
+func (m *DeviceManager) undoCreateRecoverySystem(t *state.Task, _ *tomb.Tomb) error {
+	if release.OnClassic {
+		// TODO: this may need to be lifted in the future
+		return fmt.Errorf("cannot run update gadget assets task on a classic system")
+	}
+
+	st := t.State()
+	st.Lock()
+	defer st.Unlock()
+
+	remodelCtx, err := DeviceCtx(st, t, nil)
+	if err != nil {
+		return err
+	}
+
+	setup, err := taskRecoverySystemSetup(t)
+	if err != nil {
+		return fmt.Errorf("internal error: cannot obtain recovery system setup information")
+	}
+	label := setup.Label
+
+	var undoErr error
+	if err := os.RemoveAll(setup.Directory); err != nil && !os.IsNotExist(err) {
+		t.Logf("when removing recovery system %q: %v", setup.Label, err)
+		undoErr = err
+	} else {
+		t.Logf("removed recovery system directory %v", setup.Directory)
+	}
+	for _, f := range setup.NewCommonFiles {
+		if err := os.Remove(f); err != nil && !os.IsNotExist(err) {
+			t.Logf("when removing seed file %q: %v", f, err)
+			if undoErr == nil {
+				undoErr = err
+			}
+		} else {
+			t.Logf("removed new file %q", f)
+		}
+	}
+
+	if err := boot.DropRecoverySystem(remodelCtx, label); err != nil {
+		return fmt.Errorf("cannot drop a current recovery system %q: %v", label, err)
+	}
```

Comment:
what happens with the bootenv status ?

Reply:
If we reached this place, recovery_system_status and try_recovery_system will have been cleared already, we never write to them after we managed to reboot back to run mode and inspected the status in ensure.